tpm2/test/util/file_io_test.c

```diff
@@ -94,3 +114,133 @@ void test_verifyOutputFilePath(void)
   remove("temp_file");
 }
  
+
+//----------------------------------------------------------------------------
+// test_read_bytes_from_file()
+//----------------------------------------------------------------------------
+void test_read_bytes_from_file(void)
+{
+  // Create a test data (and length) example for use in tests
+  uint8_t * testfile_data = (uint8_t *) "123 & ABC !!";
+  size_t testfile_size = strlen((char *) testfile_data);
+
+  // Create data amd data_length parameters to use in test function calls
+  uint8_t * testdata = NULL;
+  size_t testdata_len = 0;
+
+  // Trying to read from NULL input path should result in error
+  CU_ASSERT(read_bytes_from_file(NULL, &testdata, &testdata_len) == 1);
+
+  // Trying to read from non-existent file should result in error
+  CU_ASSERT(access("fake_file", F_OK) == -1);
```

Comment:
Agreed. I should have probably just gotten rid of this, and in the end I did get rid of it, but I first changed this to an if (access()) statement where I removed "fake_file" if it existed before running the read_bytes_from_file() test. Then I thought about the access() call not distinguishing between an existing file without permissions and a non-existing file, so I added a test for bad permissions case. Then I realized that if I just moved this test until after I removed the actual test file, I would know that the file didn't exist, because I just removed it. Then I realized that these file I/O tests create a lot of different test files with different names and thought it might be a good idea to standardize on something like "testfile".

Anyway, I really overthought this and have made a bunch of additional changes to this file that you may or may not disagree with. It is not a big deal to revert or change differently.

Also, in some ways I wonder if we need to retest the input/output path verification functionality in read_bytes_from_file() and write_bytes_to_file() since it is implemented in the verifyInputFilePath() and verifyOutputFilePath() functions that are called here, but thought it may not be terrible since those are the types of ways one can tweak the variables in the function's parameter list. This pattern of retesting functionality in nested functions, though, will be pretty common, so we should probably adopt a consistent viewpoint on it.

Reply:
I prefer the way you've done it, testing all the combinations for each function even though some of the checks actually happen in other functions. I don't know that we would ever do it, but it protects us against changing `read_bytes_from_file()` or `write_bytes_from_file()` to not rely on `verifyOutputFilePath()` (or changing how `verifyOutputFilePath()` acts). I think it's good practice going forward as well.